Issue/6339 react to beta switch

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #6339
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds a way to listen to preferences changes using the AppPrefsWrapper, which allows reacting to the changes and refreshing the UI, the goal here is to fix the issue of not refreshing the menu buttons when the coupon toggle is refreshed.

### Testing instructions
1. Make sure "Don't keep activities" is disabled (this was necessary to reproduce the original issue)
2. Go to Settings > Beta Features
3. Turn on Coupon Management
4. Go back to the hub menu screen
5. Confirm the coupons button is visible correctly.


- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Add a way to observe preferences

hichamboushaba committed:
Refresh menu buttons when coupon beta feature is changed

hichamboushaba committed:
Add a test case

hichamboushaba committed:
Update logic to use Unit instead of String

hichamboushaba committed:
Revert unwanted change

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppPrefsWrapper.kt:
```diff
@@ -117,4 +123,20 @@ class AppPrefsWrapper @Inject constructor() {
+    /**
+     * Returns a Flow that emits the keys of the changed preferences.
+     */
+    fun observePrefs(): Flow<String> {
+        return callbackFlow {
+            val listener = OnSharedPreferenceChangeListener { _, key ->
+                trySend(key)
+            }
+            AppPrefs.getPreferences().registerOnSharedPreferenceChangeListener(listener)
+
+            awaitClose {
+                AppPrefs.getPreferences().unregisterOnSharedPreferenceChangeListener(listener)
+            }
+        }
+    }
```
This couples the listeners to the keys, which are private in most cases (I'm not sure they should be 🤔), but I think it's OK for most cases.

We can also create different versions of the `observeXXX` functions for each preference we want to observe, but IMO it doesn't scale well.

Let me know what you think 🙏.

wzieba replied:
WDYT about rather using [`DataStore` ](https://developer.android.com/topic/libraries/architecture/datastore), a recommended successor of SharedPreferences with built-in reactiveness? There are a few benefits:

![image](https://user-images.githubusercontent.com/5845095/164913641-89cdda39-717d-4464-914a-ebf50a778b9c.png)

wzieba replied:
> We can also create different versions of the observeXXX functions for each preference we want to observe, but IMO it doesn't scale well.

IMO it's a better solution to provide separate `observeXXX` functions. In its current form, ViewModels will have to know the keys of shared preferences which are an implementation detail of the data layer. WDYT?

hichamboushaba replied:
Thanks for chiming in @wzieba .
> WDYT about rather using [DataStore](https://developer.android.com/topic/libraries/architecture/datastore)

This is a bigger project, if the team wants to migrate all the preferences to DataStore then it has to be done as part of a hack week or a separate project, but moving just a unique preference to have observability will just cause more inconsistencies IMO 🙂.

> ViewModels will have to know the keys of shared preferences which are an implementation detail of the data layer

Why do you think the keys are an implementation detail? the preferences are a store of key-value, and IMO the `key` is part of the contract, where how data is saved is the implementation detail.
Exposing the keys won't cause any tight coupling with the implementation of the data layer, we can easily migrate to another `key-value` store like the `DataStore`'s `Preferences` for example.
This reasoning is why I said the remark:
>  I'm not sure they should be 🤔

but if I'm wrong, I'd love to hear your arguments 🙂.

wzieba replied:
> This is a bigger project, if the team wants to migrate all the preferences to DataStore then it has to be done as part of a hack week or a separate project 

Gotcha, I just want to mention that we already use DataStore for saving values for WCTracker so there's a setup for that. I don't think we have to migrate all SharedPreferences at once - I believe we can do this one by one and add new values to DataStore. I think it would be great to start using DataStore 🙂 WDYT?

> Why do you think the keys are an implementation detail?

IMO ViewModel shouldn't know that the data layer uses a key-value strategy (or any other). Let's take e.g. different storage - Firebase Remote Config. If we'll decide to use it at some point (to enable/disable features remotely) and we'll want to re-use the same key as we do for local key-store values. This will end up with defining Firebase Remote Config keys in the presentation/domain layers of the Woo app, which doesn't sound right.

Another argument would be to wonder what if SharedPreferences had a different API. E.g. let's imagine that SharedPrefences use `kapt` for generating getters. Would we still use a key-value strategy? Probably not, we probably would just cover those generated methods via an interface. This would mean that SharedPreferences API indicates our domain implementation and this concludes (to me) that using keys is an implementation detail.

Also, as you say, we will be able to easily migrate to another key-value store, but what about others, non-key-value storage?

That's why I believe keys are private in `AppPrefs` and values are exposed via methods. WDYT?

hichamboushaba replied:
> I don't think we have to migrate all SharedPreferences at once

This is where we have different opinions, I think we should migrate the preferences all at once for consistency, or at least we spend enough time to design the API before the migration.
And personally, I don't think the advantages of DataStore warrants an urgent migration, SharedPreferences when not abused, still does a good job, the issue with it was for apps that used it to store large objects of data, which wasn't what it was designed for.

> IMO ViewModel shouldn't know that the data layer uses a key-value strategy

Thanks so much for the long answer, I truly appreciate it.
While I see your point, I still don't agree with it 😅:
- Most (if not all) preferences frameworks use key-value storage (NSUserDefaults in iOS, localStorage for web...), and this is why even multiplatform frameworks expose a key-value API (for example: https://github.com/florent37/Multiplatform-Preferences), which means exposing the keys doesn't add any coupling to the platform nor the framework.
- Exposing the keys allows some flexible ways for observing changes, you can design your keys in a way that allows grouping them into different categories (prefixes or suffixes), and observe a bulk of them directly.

Still, I decided to remove exposing the keys in this PR, so now the API exposes just a `Flow<Unit>`, and it's up to the client to check which values did change (similar to what `PreferencesDataStore` does), check the commit: cd69710.

wzieba replied:
> This is where we have different opinions, I think we should migrate the preferences all at once for consistency, or at least we spend enough time to design the API before the migration.

I get your point but I think it'll be difficult to do. I think that this migration can be easily done step-by-step instead all at once - the same as we didn't migrate to Kotlin all at once but step by step (starting with writing new classes in Kotlin).

> Most (if not all) preferences frameworks use key-value storage

I understand but I still stand on the point that the data layer shouldn't dictate our API, even if the majority of frameworks use key-value (or our domain/presentation dictate data-specific keys).

> Still, I decided to remove exposing the keys in this PR, so now the API exposes just a Flow<Unit>, and it's up to the client to check which values did change

I see, nice idea! I think it fixes my main concern about exposing keys but I still would vote to move it to some repository to hide it from VM and expose only `observeCouponBetaSwitch` (like `DataStoreTrackerRepository#observeLastSendingDate`).

> While I see your point, I still don't agree with it 😅:

My ego screams 😄 but it's a good thing I believe!

> Thanks so much for the long answer, I truly appreciate it.

Thank you too for the discussion! It gave me more light on how we should/could approach keeping preferences in our app. 🙌

hichamboushaba replied:
> I get your point but I think it'll be difficult to do

We don't have that many preferences, it shouldn't be that hard, I think it's manageable in a hack week if we decide the advantages of DataStore are worth the migration 🙂.

Thanks again for the discussion @wzieba 🙏

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenuViewModel.kt:
```diff
@@ -15,33 +14,31 @@ import com.woocommerce.android.analytics.AnalyticsTracker.Companion.VALUE_MORE_M
-            selectedSite.observe().filterNotNull()
-        ) { count, selectedSite ->
+            selectedSite.observe().filterNotNull(),
+            appPrefsWrapper.observePrefs().filter { it == "IS_COUPONS_ENABLED" }.map {}.onStart { emit(Unit) }
```
I didn't want to make DeletablePrefKey public (I don't know why they have to be private, but I don't want to go ahead with changing it in this PR), so I'm using the plain String value of the key, I don't like it, but I think it does the job.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenuViewModel.kt:
```diff
@@ -94,7 +91,7 @@ class MoreMenuViewModel @Inject constructor(
         }
 
-    private fun SiteModel.getSelectedSiteAbsoluteUrl(): String = url.toUri().host ?: ""
+    private fun SiteModel.getSelectedSiteAbsoluteUrl(): String = runCatching { URL(url).host }.getOrDefault("")
```
I had to resort to java.net.URL to make the class testable.

hichamboushaba committed:
Remove the coupling between preferences keys and observers

hichamboushaba committed:
Merge branch 'trunk' into issue/6339-react-to-beta-switch

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/MoreMenuViewModel.kt

hichamboushaba committed:
Refactor logic to move the observer handling to the repository

hichamboushaba commented:
@0nko FYI, I've refactored the logic a bit after the merge to fix the conflicts, as a Repository was added, I moved the logic to it, and this also aligns with Wojtek's opinion above.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/moremenu/domain/MoreMenuRepository.kt:
```diff
@@ -25,4 +30,9 @@ class MoreMenuRepository @Inject constructor(
         }
+
+    fun observeCouponBetaSwitch() = appPrefsWrapper.observePrefs()
+        .onStart { emit(Unit) }
```
Clever!

0nko reviewed (approved):
An elegant way to add reactiveness. Nice work! 👍
